src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/AbstractFlowScanner.java

```diff
@@ -0,0 +1,344 @@
+ *   <li>Each implementation provides its own iteration order (described in its javadoc comments),
+ *     but it is generally unsafe to rely on parallel branches being visited in a specific order.</li>
+ *   <li>Implementations may visit some or all points in the DAG, this should be called out in the class's javadoc comments</li>
+ *   <li>FlowScanners are NOT thread safe, for performance reasons and because it is too hard to guarantee.</li>
+ *   <li>Many fields and methods are protected: this is intentional to allow building upon the implementations for more complex analyses.</li>
+ *   <li>Each FlowScanner stores state internally for several reasons:</li>
+ *   <ul>
+ *      <li>This state can be used to construct more advanced analyses.</li>
+ *      <li>FlowScanners can be reinitialized and reused repeatedly: avoids the overheads of creating scanners repeatedly.</li>
+ *      <li>Allows for caching to be added inside a FlowScanner if desired, but caching is only useful when reused.</li>
+ *   </ul>
+ *
+ * <p/><strong>Suggested uses:</strong>
+ *   <ul>
+ *   <li>Implement a {@link FlowNodeVisitor} that collects metrics from each FlowNode visited, and call visitAll to extract the data.</li>
+ *   <li>Find all flownodes of a given type (ex: stages), using {@link #filteredNodes(Collection, Collection, Predicate)}</li>
+ *   <li>Find the first node with an Error before a specific node</li>
+ *   <li>Scan through all nodes *just* within a block
+ *      <ul>
+ *        <li>Use the {@link org.jenkinsci.plugins.workflow.graph.BlockEndNode} as the head</li>
+ *        <li>Use the {@link org.jenkinsci.plugins.workflow.graph.BlockStartNode} as its blacklist with {@link Collections#singleton(Object)}</li>
+ *     </ul></li>
+ *   </ul>
+ *
+ * @author <samvanoort@gmail.com>Sam Van Oort</samvanoort@gmail.com>
+ */
+public abstract class AbstractFlowScanner implements Iterable <FlowNode>, Filterator<FlowNode> {
+
+    protected FlowNode myCurrent;
+
+    protected FlowNode myNext;
+
+    protected Collection<FlowNode> myBlackList = Collections.EMPTY_SET;
+
+    /** When checking for blacklist membership, we convert to a hashset when checking more than this many elements */
+    protected static final int MAX_LIST_CHECK_SIZE = 5;
+
+    /** Helper: convert stop nodes to a collection that can efficiently be checked for membership, handling null if needed */
+    @Nonnull
+    protected Collection<FlowNode> convertToFastCheckable(@CheckForNull Collection<FlowNode> nodeCollection) {
+        if (nodeCollection == null || nodeCollection.size()==0) {
+            return Collections.EMPTY_SET;
+        } else  if (nodeCollection.size() == 1) {
+            return Collections.singleton(nodeCollection.iterator().next());
+        } else if (nodeCollection instanceof Set) {
+            return nodeCollection;
+        }
+        return nodeCollection.size() > MAX_LIST_CHECK_SIZE ? new HashSet<FlowNode>(nodeCollection) : nodeCollection;
```

Comment:
Should the method actually return the unmodifyableSet() ? Otherwise we may be at risk since the representation may be forwarded to external methods => there is a risk of concurrency issues if the API gets misused